### ISSUE TITLE
Add code to extract skills from CV PDF

### ISSUE THREAD

**Issue by ivan-kolmychek-devv** [DevvJobs/candidates-web#37]
This issue is a part of [work on parsing CVs PDF](https://github.com/DevvJobs/candidates-feedback/issues/2#issuecomment-474374208).

When user provides PDF with their CV right now to the app, it extracts all the text and doesn't do much with it. :slightly_smiling_face: 

We should extract skills (either from extracted text or directly from CV in the same manner as we extract all text now), put them into the state and display on the page, for debugging purposes.

From there on we can proceed with next parts of plan mentioned above. :wrench: 

As usual, feedback is welcome, on both this issue and the plan in general.

**Comment by lilsweetcaligula** [comment]
This feature would probably require a gql endpoint which returns a list of supported skills. Hard-coding skills may be not scalable.

**Comment by lilsweetcaligula** [comment]
This feature could probably use an additional gql endpoint to serve the list of available skills. We currently have the `Skills`query, but it only seems to work if you have a query.

It would otherwise require a `query Skills` request per word. And then, to make things interesting, there are some skills which have more than one word in the name.

**Comment by lilsweetcaligula** [comment]
I thought about CV parsing a bit more.

I came to a conclusion that we may not need fuzzy search after all. The reason for that is candidates usually proofread their CVs and as such the chance of a typo is minimal. While fuzzy-search may offer good recommendations and auto-complete capabilities, I believe it is of little use in this particular scenario.

To potentially simplify the parsing a little bit, I wanted to recommend that a separate table be defined on backend, if appropriate. This table can roughly be called `SkillAlias`, and have the `skill_id` and `alias` fields.

Then, a gql endpoint will have to be defined to serve aliases based on a given string (although this part I have not thought through enough yet, and it may need to be discussed).

The algorithm will then go through the list of all aliases and check if the CV includes the alias. If it does, then we assume the skill has been specified in the CV.

For example, given a skill `Ruby on Rails`, it can have 3 aliases: `Ruby on Rails` (itself as an alias), `Rails` and `RoR`.

What would you guys suggest?

**Comment by ivan-kolmychek-devv** [comment]
Sorry for the delay in the reply. We actually had quite lengthy discussion about it until we came to the agreement on this.

> This feature could probably use an additional gql endpoint to serve the list of available skills. We currently have the Skills query, but it only seems to work if you have the string to search by.
>
> It would otherwise require a query Skills request per word. And then, to make things interesting, there are some skills which have more than one word in the name.
>
> Or alternatively it would require a query Skills request per each letter in the alphabet plus digits and some misc. characters, but then I am not sure if there are limits on how many skills can be returned at a time.

Yes, we do limit results current skills query to few matches only and we do require query so we can find those matches.

We are not sure asking server for list of all possible skills is a good idea. We assume that, for the most CVs, the list of skills that are specified in CV is much smaller than the list of all the skills that we have on server.

Do we have any way to try to extract that list from CV without relying on the full skills list?

**Comment by ivan-kolmychek-devv** [comment]
> I came to a conclusion that we may not need fuzzy search after all. The reason for that is candidates usually proofread their CVs and as such the chance of a typo is minimal. While fuzzy-search may offer good recommendations and auto-complete capabilities, I believe it is of little use in this particular scenario.
>
>To potentially simplify the parsing a little bit, I wanted to recommend that a separate table be defined on backend, if appropriate. This table can roughly be called SkillAlias, and have the skill_id and alias fields.
>
>Then, a gql endpoint will have to be defined to serve aliases based on a given string (although this part I have not thought through enough yet, and it may need to be discussed).
>
>The algorithm will then go through the list of all aliases and check if the CV includes the alias. If it does, then we assume the skill has been specified in the CV.
>
>For example, given a skill Ruby on Rails, it can have 3 aliases: Ruby on Rails (itself as an alias), Rails and RoR.

I think there are two suggestions packed here - one is about having list of aliases in general and the other one is about exposing them all via GraphQL query.

Let me address first exposing them all via GraphQL query - if we assume they will be used in a similar manner as suggested "list of all skills" query, then same concern applies, but now there can be more than one entry per skill. We would prefer to explore other options first.

As for having the list of aliases in general, most of the aliases are already covered by fuzzy search. The aliases that it does not cover usually are abbreviations, like RoR, JS, CI/CD, etc. Some of them we cover by placing them in the name, like "JavaScript (JS)". Some of them we do not cover at all, like RoR.

It is a bit inconsistent and is not pretty at all, but it works for now. Unless we have a significant reason to rework them, re-designing this system is quite low on our current list of priorities. We will not get to it any time soon.

**Comment by ivan-kolmychek-devv** [comment]
Summarizing, it sounds to us that the question that frontend should ask server should not be "what are all the skills that you have?" but more of a "hey, do you have something like these few particular skills that I've got from users CV?".

As always, if you disagree with anything, have any feedback on this or any suggestions, please let us know.

**Comment by lilsweetcaligula** [comment]
@ivan-kolmychek-devv I'll reply within 10 minutes. Please standby.

**Comment by ivan-kolmychek-devv** [comment]
@lilsweetcaligula yes, that is an alternative to the server-side list. It also made an appearance in the internal discussion, both as permanent and temporary solution.

It sounds that neither you nor us want to ask server for available skills for every word on the CV, as it is neither accurate nor efficient. The approach you suggest (in a simplified manner) goes "ok, we have a list of tokens, let's try to find them in CV".

It still looks to us like it is a work-around for us not being able to look into PDF (or extracted text) and say "aha, this part looks like it has skills section, that consists of `Foo on Bar`, `Quizzical Biz` and `Baz`. We know that because section has header "My skills" (which is in our list of known headers for skills section) and each of those items are separated by a newline. We should ask server about each of these strings as they are and take the first suggested match".

We think that it is the question we should look into more - can we somehow get out of CVs this information? If not, what exactly prevents us? If yes, what are the challenges that we see emerging with it?

This way we can either a) discover a way to do that, or b) figure out that your suggestion is actually the only one that is viable, or even c) figure out that the whole setup is not usable in this manner and we should look into alternatives.

Each of them sounds like a useful discovery, at least to us. :slightly_smiling_face:

**Comment by lilsweetcaligula** [comment]
> We think that it is the question we should look into more - can we somehow get out of CVs this information? If not, what exactly prevents us? If yes, what are the challenges that we see emerging with it?

The problem with CVs is that there's a hundred thousand different ways to format and structure them, and any assumptions can be easily broken. Guessing is hard...

On top of the unpredictability, people would probably want to upload their CV in Spanish or French, making it impossible to use any English-based heuristics. (but it would probably be reasonable to stick to English for now, so this should not be a problem)

Example 1:
```
Test van Testen
Ruby, Rails, Postgres
```

Example 2:
```
Test van Tastisch
Skills:
- Ruby
- Rails
- Postgres
```

Example 3:
```
Test van Tastisch,
... a Rails developer with 30 years of experience, spanning a wide range of technologies, such as Postgres, JavaScript and Java.
```

Example 4:
```
Test van Tastisch
Rails | Ruby | Postgres | Java
```

We could alternatively make certain assumptions, check them, and then fall back to brute-force, but then it feels more like a set of optimizations.

Another problem with parsing PDFs is shenanigans like this one, making it difficult to rely on formatting in a PDF file:
https://github.com/DevvJobs/candidates-web/issues/45

So currently I see no viable alternative solution, except for fuzzy-searching the content of a CV and matching against a set of skills.

**Comment by ivan-kolmychek-devv** [comment]
@lilsweetcaligula speaking for both me and @tetiana-kolmychek-devv, we discussed it more and we can see your point.

The end-goal is to make it easier to fill in the data by providing a CV. This includes skills and also years of experience for them, whenever possible.

We also assume that having false-negatives is more acceptable than false-positives. We think that this feature will be of most use for people with big list of skills (10+), people with 2-3 skills can easily (and probably faster) fill them in by hand.

We had some specific ideas on how to do that, but since it's not any of us who is implementing it, they are mere suggestions. As I have mentioned before, we think that anyone who actually goes and 
spends time on this will know better than us. :slightly_smiling_face: 

Given all that, we have few concerns with parsing whole document in the suggested manner instead of trying to detect structure:

- Potential for false-positives. A lot of parts of CV can include words that are similar enough to skills to be detected as such, like company names, hobbies and interests, some words in general descriptions and such. We acknowledge that it can be handled by improving parsing.

- Years of experience. While we did plan them to be included, as they are a required information for the skill, we do not know yet how would we suggest to detect them and is it even possible. We do not 
see how it can fit into current approach, but it's out of scope for this particular task and suggestions are welcome.

Both these things add time that user needs to spend to work through the result of parsing before we run the import.

This is why we suggested earlier in this issue to look into detecting structural parts with skills and rolling from there. Our thought was that, if it's hard to do in the text, maybe we should look into alternatives, like parsing PDF into something else before.

But as long as it works and is acceptable to users, we do not really  mind it being done in another way. :slightly_smiling_face: 

So it's up to you.

_(We plan to also post reply in PR with more stuff that is specific to current implementation. Stay tuned. :slightly_smiling_face:)_

**Comment by ivan-kolmychek-devv** [comment]
@lilsweetcaligula [wrote](https://github.com/DevvJobs/candidates-web/pull/44#issuecomment-476592313) that he does not have time anymore for this issue.

If anyone wants to pick it up or contribute in any other way - go ahead. :+1:
